rfcs/049-catalogue-api-aggregations-modelling.md

```diff
@@ -0,0 +1,176 @@
+The API retrieves the `display` field, and treats it as opaque JSON -- without any knowledge of its structure.
+
+We'd hoped this would allow us to remove the internal/display models from the API repos, but the current implementation of aggregations make this tricky.
+
+This RFC proposes a change to how aggregations are handled that should remove this obstacle.
+
+[RFC 047]: ./047-catalogue-api-index-structure.md
+
+
+
+## Context: the desired behaviour
+
+Clients of the catalogue API can request aggregations of certain values (e.g. languages, licenses, work types).
+
+These are presented as a list of `AggregationBucket` in the API responses, i.e.:
+
+```
+AggregationBucket[T] {
+  data: T
+  count: Int
+  type: String
+}
+```
+
+where `T` is the same as the value appearing in the Work model.
+
+For example, languages on a work appear as follows:
+
+```
+{
+  "languages": [
+    {
+      "id": "eng",
+      "label": "English",
+      "type": "Language"
+    },
+    {
+      "id": "fre",
+      "label": "French",
+      "type": "Language"
+    }
+  ],
+  ...
+}
+```
+
+and we see that `Language` type in the aggregation buckets:
+
+```
+"aggregations": {
+  "languages": {
+    "buckets": [
+      {
+        "data": {
+          "id": "eng",
+          "label": "English",
+          "type": "Language"
+        },
+        "count": 691840,
+        "type": "AggregationBucket"
+      },
+      {
+        "data": {
+          "id": "fre",
+          "label": "French",
+          "type": "Language"
+        },
+        "count": 67187,
+        "type": "AggregationBucket"
+      },
+...
+```
+
+This presents a clean, consistent interface to clients – a value looks the same whether it's in a work or in an aggregation.
+
+We don't to change this behaviour.
+
+
+
+## The current implementation, and the problems it poses
+
+The API uses [Elasticsearch aggregations][es_aggs] to aggregate over a field in internal model.
+Elasticsearch will return single string values, which the API then serialises into the display model.
+
+For example, our language aggregation starts as [an ES terms aggregation][terms_agg] over the `data.languages.id` field:
+
+```http
+GET /works-indexed-2022-04-28/_search
+{
+  "aggs": {
+    "languages": {
+      "terms": { "field": "data.languages.id" }
+    }
+  }
+}
+```
+
+The Elasticsearch API response can only use a single string in its buckets, for example:
+
+```
+{
+  "aggregations" : {
+    "languages" : {
+      "buckets" : [
+        {
+          "key" : "eng",
+          "doc_count" : 691840
+        },
+        {
+          "key" : "fre",
+          "doc_count" : 67187
+        },
+        ...
+```
+
+The API has to contain enough internal/display model logic to interpret these values -- to know that, say, `eng` means English and it's serialised as id/label/type.
+This is precisely the sort of model coupling we're trying to get away from.
```

Comment:
For a public standard (like languages), I don't see this problem as being one of model coupling in the way formats are.  Having code to translate from `ger` to "German" on both sides makes some sense to me.  However, as a general purpose rule (e.g. complex types, below), and also for efficiency/simplicity at the front end, it is a good idea to return ids and labels together.